WP-r55657: Users: Cache database queries within WP_User_Query class

mattyrob opened the pull request:
## Description
 Users: Cache database queries within WP_User_Query class.

Cache the results of database queries within WP_User_Query class. Only cache queries that are requesting 3 or less fields so that caches are not storing full user objects. Cache results are stored in a new global cache group named users-queries. Add a new parameter to WP_User_Query called cache_results to allow developers to opt out of a receiving cached results. cache_results parameter defaults to true. Also add a new helper function called wp_cache_set_users_last_changed, similar to wp_cache_set_posts_last_changed that incroments last changed value in cache group users. Ensure that wp_cache_set_users_last_changed is called whenever user / user meta is modified for proper cache invalidation.

Props johnjamesjacoby, spacedmonkey, westi, dd32, strategio, srikanthmeenakshi, OllieJones, khoipro, rjasdfiii, flixos90, mukesh27, peterwilsoncc. 

https://core.trac.wordpress.org/changeset/55657

## Motivation and context
Backport of upstream efficiency enhancement

## How has this been tested?
This is a backport, local tests have been passing.

## Screenshots
N/A

## Types of changes
- Code efficiency enhancement

spacedmonkey committed:
WP-r55657: Users: Cache database queries within `WP_User_Query` class.

Cache the results of database queries within `WP_User_Query` class. Only cache queries that are requesting 3 or less fields so that caches are not storing full user objects. Cache results are stored in a new global cache group named `users-queries`. Add a new parameter to `WP_User_Query` called `cache_results` to allow developers to opt out of a receiving cached results. `cache_results` parameter defaults to true. Also add a new helper function called `wp_cache_set_users_last_changed`, similar to `wp_cache_set_posts_last_changed` that incroments last changed value in cache group `users`.  Ensure that `wp_cache_set_users_last_changed` is called whenever user / user meta is modified for proper cache invalidation.

WP:Props johnjamesjacoby, spacedmonkey, westi, dd32, strategio, srikanthmeenakshi, OllieJones, khoipro, rjasdfiii, flixos90, mukesh27, peterwilsoncc.
Fixes https://core.trac.wordpress.org/ticket/40613.

Conflicts:
- src/wp-includes/user.php

---

Merges https://core.trac.wordpress.org/changeset/55657 / WordPress/wordpress-develop@19bd759db5 to ClassicPress.

mattyrob committed:
Fix merge conflicts and layout changes from 55657

spacedmonkey committed:
WP-r55680: Users: Change cache group from `users-queries` to `user-queries`.

The cache group `users-queries` was added in https://core.trac.wordpress.org/changeset/55657. This global cache group, was named to be inline with cache groups added in https://core.trac.wordpress.org/changeset/55526. However, the naming of the group does not match, as other cache groups, do not end with s at the end. This change fix this naming.

WP:Props spacedmonkey, SergeyBiryukov, peterwilsoncc.
See https://core.trac.wordpress.org/ticket/40613.

---

Merges https://core.trac.wordpress.org/changeset/55680 / WordPress/wordpress-develop@201a30d387 to ClassicPress.

mattyrob committed:
Fix coding standards spacing in test

xxsimoxx commented:
Running unit tests on my MAMP stack:

```
There were 2 errors:

1) Tests_Admin_WpAutomaticUpdater::test_is_allowed_dir_should_return_true_if_open_basedir_is_set_and_path_is_allowed
PHPUnit\Framework\Exception: PHP Warning:  file_put_contents(): open_basedir restriction in effect. File(/private/var/folders/qw/yd47hwjd0dz1c0b6582l_8840000gn/T/phpunit_Mcg0xl) is not within the allowed path(s): (/Users/simo/Sites/) in Standard input code on line 92
PHP Warning:  file_put_contents(/private/var/folders/qw/yd47hwjd0dz1c0b6582l_8840000gn/T/phpunit_Mcg0xl): Failed to open stream: Operation not permitted in Standard input code on line 92

/Users/simo/Sites/ClassicPress-v2/vendor/bin/phpunit:123

2) Tests_Admin_WpAutomaticUpdater::test_is_allowed_dir_should_return_false_if_open_basedir_is_set_and_path_is_not_allowed
PHPUnit\Framework\Exception: PHP Warning:  file_put_contents(): open_basedir restriction in effect. File(/private/var/folders/qw/yd47hwjd0dz1c0b6582l_8840000gn/T/phpunit_z4FPa2) is not within the allowed path(s): (/Users/simo/Sites/) in Standard input code on line 92
PHP Warning:  file_put_contents(/private/var/folders/qw/yd47hwjd0dz1c0b6582l_8840000gn/T/phpunit_z4FPa2): Failed to open stream: Operation not permitted in Standard input code on line 92

/Users/simo/Sites/ClassicPress-v2/vendor/bin/phpunit:123
```

mattyrob commented:
These issues are fixed in #218 already but I made this backport about 3 months ago and that test fix was added about 5 weeks ago.
The branch could be sync'd to `develop` or we can ignore these local issues.

xxsimoxx reviewed (approved):
LGTM.

SergeyBiryukov committed:
WP-r56381: Users: Correct the logic for `switch_to_blog()` in `WP_User_Query::generate_cache_key()`.

If `$blog_id` equals `0`, it should be treated as the current site ID, and there is no need to switch to a different site.

This commit prevents an unnecessary call to `switch_to_blog()` on single site to avoid a fatal error when using `'orderby' => 'post_count'` and the deprecated `'who' => 'authors'` parameter:
{{{
Uncaught Error: Call to undefined function switch_to_blog() in wp-includes/class-wp-user-query.php:1077
}}}

Follow-up to https://core.trac.wordpress.org/changeset/55657.

WP:Props dd32, austinginder, RavanH, mukesh27.
Fixes https://core.trac.wordpress.org/ticket/59011.

---

Merges https://core.trac.wordpress.org/changeset/56381 / WordPress/wordpress-develop@e5b586bfd0 to ClassicPress.

viktorix reviewed (approved):
Looks good.

mattyrob commented:
Has two approvals, merging in preparation for v2 repository migration.
